LoginToTutByTest and LogOutTutByTest classes are created based on Page Object pattern.

LenousikK opened the pull request.

LenousikK committed:
LoginToTutByTest and LogOutTutByTest classes are created based on Page Object pattern.

EugenBorisik commented on src/test/java/BeforeAfter.java:
```diff
@@ -0,0 +1,23 @@
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class BeforeAfter {
+    public WebDriver driver;
```
Lines 9-10: Why are they public?

LenousikK replied:
It's my bad. I just didn't pay attention.
Now it is corrected.

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,88 @@
+import static org.openqa.selenium.support.ui.ExpectedConditions.titleIs;
+
+public class LoginPage extends PageBasis {
+    private static final String URL = "https://www.tut.by/";
```
Optional: it is more readable as for me, when fields are grouped by types

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,88 @@
+        driver.get(URL);
+    }
+
+    public String loginTitle() {
```
Move this variable in the class, in which you need it

LenousikK replied:
I removed this method and remained only the variable.

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,88 @@
+        return LOGIN_TITLE;
+    }
+
+    public WebElement loginMenuItem() {
```
Page object encapsulates all actions on the page. It means that PO provides just methods like: login(...), logout(...), create(...). In your solution I see many public methods returining WebElement and By - it is mistake. Please update code.

LenousikK replied:
I updated access modifiers for methods depending on the place where they are called.

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,88 @@
+        return buttonLoginSubmit;
+    }
+
+    public void login() {
```
login() method must have 2 parameters: string for username and string for password. Otherwise your method is not flexible and it is possible to login only with one set of credentials

LenousikK replied:
I agree and updated this.

EugenBorisik commented on src/test/java/LoginToTutByTest.java:
```diff
@@ -0,0 +1,14 @@
+    public void loginSuccessful() {
+        LoginPage loginPage = new LoginPage(driver, wait);
+        loginPage.login();
+        wait.until(ExpectedConditions.presenceOfElementLocated(loginPage.labelLoggedUsername()));
```
This waiter must be placed in login() method.

LenousikK replied:
Corrected.

EugenBorisik commented on src/test/java/PageBasis.java:
```diff
@@ -0,0 +1,11 @@
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class PageBasis {
+    public WebDriver driver;
```
Why are fields public?

LenousikK replied:
Same as before - just missed this moment.

EugenBorisik commented on src/test/java/LogOutTutByTest.java:
```diff
@@ -0,0 +1,14 @@
+        loginPage.login();
+        loginPage.linkLogoutMenuItem().click();
+        loginPage.buttonLogout().click();
+        assertEquals("Войти", loginPage.getTextOfLabelLoggedOut());
```
Move your string in constant

LenousikK replied:
Corrected.

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,88 @@
+        return buttonLoginSubmit;
+    }
+
+    public void login() {
```
One more moment: header changes after login. Let's create one more page object class - LoggedInPage. And your login method will return LoggedInPage instance. Signature will look like - public LoggedInPage login(...)
If you have any questions - feel free to ask

LenousikK replied:
A new class is created. The method is re-written as required.

EugenBorisik reviewed (changes requested):
@LenousikK please update code according to comments + update the same places in the second PR

LenousikK committed:
Applied corrections:
- WebDriver and WebDriverWait variables are declared as protected;
- methods implemented for the page are public;
- a new class LoggedInPage is created;
- loginPage.login() method is re-written with a new signature (it returns LoggedInPage type and login/ passwords are passed as parameters) and some changes in its realization (in particular, login/ password are not hard-coded anymore, wait.until(ExpectedConditions.presenceOfElementLocated(labelLoggedUsername()) is added inside this method);
- Expected text in assertions is moved out to class variables.

LenousikK committed:
Update access modifiers from public to private for the following methods:
- enterLoginUsername();
- enterLoginPassword().

EugenBorisik commented on src/test/java/LoggedInPage.java:
```diff
@@ -0,0 +1,33 @@
+        WebElement buttonLogout = driver.findElement(BUTTON_LOGOUT);
+        return buttonLogout;
+    }
+    public void logout(){
```
After logout you are on the LoginPage, so return it

LenousikK replied:
Corrected

EugenBorisik commented on src/test/java/LoggedInPage.java:
```diff
@@ -0,0 +1,33 @@
+        buttonLogout().click();
+    }
+
+    public String getTextOfLabelLoggedOut() {
```
Make sense to move into LoginPage

LenousikK replied:
Corrected

EugenBorisik commented on src/test/java/LoginPage.java:
```diff
@@ -0,0 +1,67 @@
+        return LABEL_LOGGED_USER_NAME;
+    }
+
+    public String getTextOfLabelLoggedUsername() {
```
Make sense to move into LoggedInPage :)

LenousikK replied:
Corrected

EugenBorisik reviewed (approved):
@LenousikK I close task, but please make updates according to additional comments.
This task is very important, that's why I'm meticulous. I even can say that you are ready to write auto-tests anywhere, if really understand page object pattern.

LenousikK committed:
Applied corrections:
>>> After logout you are on the LoginPage, so return it
Now logout() method is re-written and returns LoginPage;

>>> Make sense to move into LoginPage
getTextOfLabelLoggedOut() method is moved to LoginPage class

>>> Make sense to move into LoggedInPage :)
getTextOfLabelLoggedUsername() method is moved to LoggedInPage class
